core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -287,6 +296,7 @@ public void runInContext() {
     }
     stream.setDecompressorRegistry(decompressorRegistry);
     channelCallsTracer.reportCallStarted();
+    this.observer = observer;
```

Comment:
You could instead pass the observer to `startDeadlineNotifyApplicationTimer` so that the timer keeps a reference to the observer. Passing the object as argument is better than saving it as a mutable field, because it's fewer states.